action/protocol/staking/ethabi/stake_base.go

```diff
@@ -1,203 +1,25 @@
-	}
-	result.SelfStakeBucketIdx = candidate.SelfStakeBucketIdx
-	if amount, ok := new(big.Int).SetString(candidate.SelfStakingTokens, 10); ok {
-		result.SelfStakingTokens = amount
-	} else {
-		return nil, errConvertBigNumber
-	}
-	return result, nil
-}
-
-func encodeBucketTypeListToEth(outputs abi.Arguments, bucketTypes *iotextypes.ContractStakingBucketTypeList) (string, error) {
-	args := make([]BucketTypeEth, len(bucketTypes.BucketTypes))
-	for i, bt := range bucketTypes.BucketTypes {
-		args[i] = BucketTypeEth{}
-		if amount, ok := new(big.Int).SetString(bt.StakedAmount, 10); ok {
-			args[i].StakedAmount = amount
-		} else {
-			return "", errConvertBigNumber
-		}
-		args[i].StakedDuration = bt.StakedDuration
-	}
-
-	data, err := outputs.Pack(args)
-	if err != nil {
-		return "", err
-	}
-	return hex.EncodeToString(data), nil
-}
-
 // BuildReadStateRequest decode eth_call data to StateContext
 func BuildReadStateRequest(data []byte) (protocol.StateContext, error) {
 	if len(data) < 4 {
-		return nil, errInvalidCallData
+		return nil, stakingComm.ErrInvalidCallData
 	}
 
-	switch methodSig := hex.EncodeToString(data[:4]); methodSig {
-	case hex.EncodeToString(_bucketsMethod.ID):
-		return newBucketsStateContext(data[4:])
-	case hex.EncodeToString(_bucketsByCandidateMethod.ID):
-		return newBucketsByCandidateStateContext(data[4:])
-	case hex.EncodeToString(_bucketsByIndexesMethod.ID):
-		return newBucketsByIndexesStateContext(data[4:])
-	case hex.EncodeToString(_bucketsByVoterMethod.ID):
-		return newBucketsByVoterStateContext(data[4:])
-	case hex.EncodeToString(_bucketsCountMethod.ID):
-		return newBucketsCountStateContext()
-	case hex.EncodeToString(_candidatesMethod.ID):
-		return newCandidatesStateContext(data[4:])
-	case hex.EncodeToString(_candidateByNameMethod.ID):
-		return newCandidateByNameStateContext(data[4:])
-	case hex.EncodeToString(_candidateByAddressMethod.ID):
-		return newCandidateByAddressStateContext(data[4:])
-	case hex.EncodeToString(_totalStakingAmountMethod.ID):
-		return newTotalStakingAmountContext()
-	case hex.EncodeToString(_compositeBucketsMethod.ID):
-		return newCompositeBucketsStateContext(data[4:])
-	case hex.EncodeToString(_compositeBucketsByCandidateMethod.ID):
-		return newCompositeBucketsByCandidateStateContext(data[4:])
-	case hex.EncodeToString(_compositeBucketsByIndexesMethod.ID):
-		return newCompositeBucketsByIndexesStateContext(data[4:])
-	case hex.EncodeToString(_compositeBucketsByVoterMethod.ID):
-		return newCompositeBucketsByVoterStateContext(data[4:])
-	case hex.EncodeToString(_compositeBucketsCountMethod.ID):
-		return newCompositeBucketsCountStateContext()
-	case hex.EncodeToString(_compositeTotalStakingAmountMethod.ID):
-		return newCompositeTotalStakingAmountContext()
-	case hex.EncodeToString(_contractBucketTypesMethod.ID):
-		return newContractBucketTypesStateContext(data[4:])
-	default:
-		return nil, errInvalidCallSig
+	methodSig, err := v1.BuildReadStateRequest(data)
```

Comment:
why v1 takes precedence over v2?

Reply:
In principle, to maintain forward compatibility, each version should have completely new interfaces without interface overlapping.